### DISCUSSION TITLE
Training StyleGAN2 in Jax/Flax

### DISCUSSION THREAD

**Discussion by matthias-wright** [jax-ml/jax#7905] [Show and tell]
Hi all,

Here is an [implementation](https://github.com/matthias-wright/flaxmodels/tree/main/flaxmodels/stylegan2) of StyleGAN2 in Jax/Flax that is compatible with the official weights. The training code can be found [here](https://github.com/matthias-wright/flaxmodels/tree/main/training/stylegan2).

**Highlights**
* Multi GPU training (data parallelism) with `jax.pmap`
* Results match the original implementation (see below)
* `tf.data.TFRecordDataset` is used for efficient data loading

**Lowlights**
* I only trained on resolution 512x512 due to limited resources
* Mixed-precision training is not stable yet. I used `flax.optim.DynamicScale` and also all the tricks from the original implementation (casting to float32 for some operations, using pre-normalization in the modulated conv layer, only using float16 for the higher resolutions, clipping the output of the convolution layers, etc) but it is not properly working yet.

I also added instructions on how to train on your own dataset. If you have any questions feel free to ask them here or on the [github](https://github.com/matthias-wright/flaxmodels) page.

Cheers!

**Some results**
![alt text](https://raw.githubusercontent.com/matthias-wright/flaxmodels/main/training/stylegan2/images/ffhq_overview.jpg)

**Style mixing**
![alt text](https://raw.githubusercontent.com/matthias-wright/flaxmodels/main/training/stylegan2/images/ffhq_style_mixing.jpg)

**Comment by 8bitmp3** [comment]
🔥

@levskaya @andsteing @marcvanzee @jheek @gnecula @avital @mattjj
